include/llbuild/Core/BuildEngine.h

```diff
@@ -204,11 +214,10 @@ class Rule {
   /// state managed externally to the build engine. For example, a rule which
   /// computes something on the file system may use this to verify that the
   /// computed output has not changed since it was built.
-  std::function<bool(BuildEngine&, const Rule&,
-                     const ValueType&)> isResultValid;
+  virtual bool isResultValid(BuildEngine&, const ValueType&);
```

Comment:
Should we make these required, just so that this is a purely abstract base class? If so, remember to give the vtable a home.

Reply:
Possibly.  The `isResultValid` method does feel like it is something that rules should be explicit about.   The `updateStatus` method, on the other hand, is definitely optional, and as I look at it perhaps could/should be removed altogether with some more refactoring.